sqllogictests: panic comparing literal and column

This query:
```sql
SELECT e
  FROM t1
 WHERE b>c
   AND d>e
 ORDER BY 1
```
produces the following error:
```
panic: interface conversion: interface {} is int64, not int32 [recovered]
	panic: interface conversion: interface {} is int64, not int32

goroutine 140 [running]:
testing.tRunner.func1(0xc4204e70e0)
	/usr/lib/go-1.10/src/testing/testing.go:742 +0x29d
panic(0x85c200, 0xc4204b7a00)
	/usr/lib/go-1.10/src/runtime/panic.go:505 +0x229
gopkg.in/src-d/go-mysql-server.v0/sql.numberT.Compare(0xc400000107, 0x82d7a0, 0xc420025120, 0x82d7a0, 0xc420025110, 0x9418c0, 0xc420024f3c, 0x1)
	/home/antonio/work/src/gopkg.in/src-d/go-mysql-server.v0/sql/type.go:264 +0x455
gopkg.in/src-d/go-mysql-server.v0/sql/expression.(*comparison).Compare(0xc4201a3dd0, 0xc42015c780, 0xc420224af0, 0x5, 0x5, 0xc420224af0, 0xc4201bbba8, 0x5c520f)
	/home/antonio/work/src/gopkg.in/src-d/go-mysql-server.v0/sql/expression/comparison.go:45 +0x2d4
gopkg.in/src-d/go-mysql-server.v0/sql/expression.(*GreaterThan).Eval(0xc4201a3dd0, 0xc42015c780, 0xc420224af0, 0x5, 0x5, 0x5, 0x5, 0xc420224af0, 0x5)
	/home/antonio/work/src/gopkg.in/src-d/go-mysql-server.v0/sql/expression/comparison.go:265 +0x5c
gopkg.in/src-d/go-mysql-server.v0/sql/expression.(*And).Eval(0xc42015dfc0, 0xc42015c780, 0xc420224af0, 0x5, 0x5, 0x0, 0x73f290, 0xc4201bbc01, 0x412898)
	/home/antonio/work/src/gopkg.in/src-d/go-mysql-server.v0/sql/expression/logic.go:46 +0x62
gopkg.in/src-d/go-mysql-server.v0/sql/plan.(*FilterIter).Next(0xc4200a7b00, 0x462, 0x73f290, 0xc4201bbd68, 0x4e74d1, 0xc4204d2d80)
	/home/antonio/work/src/gopkg.in/src-d/go-mysql-server.v0/sql/plan/filter.go:94 +0x97
gopkg.in/src-d/go-mysql-server.v0/sql/plan.(*sortIter).computeSortedRows(0xc4204b79c0, 0xc4201bbd88, 0x4cffca13)
	/home/antonio/work/src/gopkg.in/src-d/go-mysql-server.v0/sql/plan/sort.go:182 +0x6d
gopkg.in/src-d/go-mysql-server.v0/sql/plan.(*sortIter).Next(0xc4204b79c0, 0xc4201bbd78, 0xc4201bbd68, 0x464c4e, 0xc4204e70e0, 0x47)
	/home/antonio/work/src/gopkg.in/src-d/go-mysql-server.v0/sql/plan/sort.go:160 +0xc1
gopkg.in/src-d/go-mysql-server.v0/sql/plan.(*iter).Next(0xc420246020, 0x7fed77a0b0d0, 0xc4201bbe08, 0x73f3d1, 0xc4204e70e0, 0x7fed77a0b0b0)
	/home/antonio/work/src/gopkg.in/src-d/go-mysql-server.v0/sql/plan/project.go:105 +0x38
gopkg.in/src-d/go-mysql-server.v0/logic.(*hashRowResult).CheckResult(0xc42015d320, 0x93c320, 0xc420246020, 0xc4204b8250, 0x1, 0x1, 0x1, 0x93c320)
	/home/antonio/work/src/gopkg.in/src-d/go-mysql-server.v0/logic/parser.go:122 +0x288
gopkg.in/src-d/go-mysql-server.v0/logic.execTestFile.func1.2(0xc4204e70e0)
	/home/antonio/work/src/gopkg.in/src-d/go-mysql-server.v0/logic/logic_test.go:96 +0x1f4
```

--- a/sql/type.go
+++ b/sql/type.go
@@ -255,39 +255,50 @@ func (t numberT) Convert(v interface{}) (interface{}, error) {
 
 // Compare implements Type interface.
 func (t numberT) Compare(a interface{}, b interface{}) (int, error) {
-	if a == b {
+	if IsUnsigned(t) {
+		return compareUnsigned(a, b)
+	}
+
+	return compareSigned(a, b)
+}
+
+func compareSigned(a interface{}, b interface{}) (int, error) {
+	ca, err := cast.ToInt64E(a)
+	if err != nil {
+		return 0, err
+	}
+	cb, err := cast.ToInt64E(b)
+	if err != nil {
+		return 0, err
+	}
+
+	if ca == cb {
 		return 0, nil
 	}
 
-	switch t.t {
-	case sqltypes.Int32:
-		if a.(int32) < b.(int32) {
-			return -1, nil
-		}
-	case sqltypes.Int64:
-		if a.(int64) < b.(int64) {
-			return -1, nil
-		}
-	case sqltypes.Uint32:
-		if a.(uint32) < b.(uint32) {
-			return -1, nil
-		}
-	case sqltypes.Uint64:
-		if a.(uint64) < b.(uint64) {
-			return -1, nil
-		}
-	case sqltypes.Float32:
-		if a.(float32) < b.(float32) {
-			return -1, nil
-		}
-	case sqltypes.Float64:
-		if a.(float64) < b.(float64) {
-			return -1, nil
-		}
-	default:
-		if cast.ToInt64(a) < cast.ToInt64(b) {
-			return -1, nil
-		}
+	if ca < cb {
+		return -1, nil
+	}
+
+	return +1, nil
+}
+
+func compareUnsigned(a interface{}, b interface{}) (int, error) {
+	ca, err := cast.ToUint64E(a)
+	if err != nil {
+		return 0, err
+	}
+	cb, err := cast.ToUint64E(b)
+	if err != nil {
+		return 0, err
+	}
+
+	if ca == cb {
+		return 0, nil
+	}
+
+	if ca < cb {
+		return -1, nil
 	}
 
 	return +1, nil

--- a/sql/type_test.go
+++ b/sql/type_test.go
@@ -29,6 +29,20 @@ func TestInt32(t *testing.T) {
 	gt(t, Int32, int32(3), int32(2))
 }
 
+func TestNumberComparison(t *testing.T) {
+	eq(t, Int64, int32(1), int32(1))
+	eq(t, Int64, int32(1), int64(1))
+	gt(t, Int64, int64(5), int32(1))
+	gt(t, Int64, int32(5), int64(1))
+	lt(t, Int32, int64(1), int32(5))
+
+	eq(t, Uint32, int32(1), uint32(1))
+	eq(t, Uint32, int32(1), int64(1))
+	gt(t, Uint32, int64(5), uint32(1))
+	gt(t, Uint32, uint32(5), int64(1))
+	lt(t, Uint32, uint64(1), int32(5))
+}
+
 func TestInt64(t *testing.T) {
 	convert(t, Int64, int32(1), int64(1))
 	convert(t, Int64, 1, int64(1))

